packages/@n8n/db/src/repositories/workflow-dependency.repository.ts

```diff
@@ -0,0 +1,80 @@
+import { Service } from '@n8n/di';
+import { DataSource, LessThan, Repository } from '@n8n/typeorm';
+
+import { WorkflowDependency } from '../entities';
+
+@Service()
+export class WorkflowDependencyRepository extends Repository<WorkflowDependency> {
+	constructor(dataSource: DataSource) {
+		super(WorkflowDependency, dataSource.manager);
+	}
+	/**
+	 * Replace the dependencies for a given workflow.
+	 * Uses the workflowVersionId to ensure consistency between the workflow and dependency tables.
+	 * @param workflowId the id of the workflow
+	 * @param workflowVersionId the version counter of the workflow
+	 * @param dependencies the new dependencies to be written
+	 * @returns whether the update was applied
+	 */
+	async updateDependenciesForWorkflow(
+		workflowId: string,
+		dependencies: WorkflowDependency[],
+	): Promise<boolean> {
+		// Handle empty dependencies (cleanup scenario)
+		if (dependencies.length === 0) {
+			return await this.manager.transaction(async (tx) => {
+				const deleteResult = await tx.delete(WorkflowDependency, { workflowId });
+				return (
+					deleteResult.affected !== undefined &&
+					deleteResult.affected !== null &&
+					deleteResult.affected > 0
+				);
+			});
+		}
+
+		const incomingVersionId = this.validateDependencies(dependencies);
+		return await this.manager.transaction(async (tx) => {
+			// Delete only dependencies with older versions
+			const deleteResult = await tx.delete(WorkflowDependency, {
+				workflowId,
+				workflowVersionId: LessThan(incomingVersionId),
+			});
+
+			// If we deleted something, the incoming version is newer - proceed with insert
+			if (deleteResult.affected && deleteResult.affected > 0) {
+				const entities = dependencies.map((dep) => this.create(dep));
+				await tx.insert(WorkflowDependency, entities);
+				return true;
+			}
+
+			// Nothing was deleted - either no existing data, or existing data is newer/same version
+			// Check if any dependencies exist for this workflow
+			const existingCount = await tx.count(WorkflowDependency, { where: { workflowId } });
+
+			if (existingCount === 0) {
```

Comment:
I think there's a race condition here, if two instances end up here at around the same time and the index is empty. Both delete nothing, both check nothing in DB --> all the dependencies will be inserted twice